### DISCUSSION TITLE
Using NVK_Create with Playtime

### DISCUSSION THREAD

**Discussion by sdfalk** [helgoboss/helgobox#1147] [Ideas]
https://github.com/user-attachments/assets/4829625e-bea7-4a48-9656-9796464897b5

So I'm wondering if I should talk to the developer of playtime, NVK, or both.
I'd like to be able to drag and drop from NVK_create into playtime, and it doesn't appear to 
work.
Might there be an update for that functionality to work at some point?

thanks

**Comment by helgoboss** [comment]
NVK is some ReaScript, right? Not sure how file dragging is done in ReaScript. Playtime supports file drag and drop from media explorer and normal file managers. I think ReaScript doesn't have real file dragging. E.g. the Gobal Sampler script is also not able to drag stuff to the file explorer (and therefore also not to Playtime). I will ask Justin what's necessary to receive ReaScript drags.

**Reply by sdfalk** [reply]
Thanks..appreciate the fast response.. If not, such is life..fingers crossed :)

**Reply by joeyxsam** [reply]
Cool request - no idea if it's possible, but you can drag from nvk_create straight to arrange window in reaper, so it seems like it would be?  Also, for whatever it's worth, I think nvk_create relies on media explorer databases in order to function properly.  I did at one time ask Nick (NVK creator) if it was possible to drag and drop multiple selected files from NVK create to a regular folder on the desktop and I believe he said it was not.

**Reply by sdfalk** [reply]
Well thats to bad..Well I'll keep my fingers crossed regardless..I guess right click on file in nvk_create and show in finder and then drag and drop is the obvious workaround..

**Comment by sdfalk** [comment]
From the developer of NVK:

**_Hi Stephen,

For that to work, we'd probably both have to pass along some data to each other. There isn't a standard way to do this in Reaper. If you get in contact with the creator of Playtime and they are open to the idea, I'd be happy to work with them to add this feature.

Cheers,
Nick**

Well he's certainly interested.._

Stephen

**Reply by helgoboss** [reply]
@nickvonkaenel Hi Nick. Open to ideas :) I'm not familiar with drag and drop in ReaScript. Could you outline how that works? I guess I would need to expose a ReaScript function from my extension?

@Bird-Bird While we are at it, I would love to add support for dragging samples in from Global Sampler into Playtime as well.

Context: Playtime is a clip launcher for REAPER (see screenshot below) and this request is about dragging material from ReaScripts such as NVK_Create or Global Sampler into Playtime. I thought it would just work until I realized that none of the ReaScripts actually use file drag-and-drop. Maybe we could find a common solution that works both for Global Sampler and NVK_Create.

![](https://www.helgoboss.org/projects/playtime/screenshots/playtime-neutral-dark.png)

**Reply by nickvonkaenel** [reply]
Hi Benjamin!

I think you are correct, unless Justin adds some kind of API for it, I'm not sure of a way to do real file dragging from ReaScript. My thought was that we could agree on an ExtState key that gets set by my script which contains the file path of a dragged file and if the mouse is released while hovering over your script window you grab that file path and maybe set a ExtState key to let my script know that the file has been added.

**Reply by helgoboss** [reply]
Thanks Nick for chiming in! Sounds like a good idea. 

With that solution, it would probably make sense to name that ExtState according to your script, so that the key starts with "NVK_Create_", for example. Conceptually, that would make NVK_Create the provider of an API that other applications can use. And Playtime would be the first consumer of that API (there could also be other consumers). What do you think?

I'll have to do some tests on macOS because I'm not sure now if an inactive window receives mouse-up events, but there must be a way.

**Reply by nickvonkaenel** [reply]
Sure. I suppose another way might be for me to only set the ExtState for a single defer cycle when the item is dragged and released over an unidentified window. Then in theory you could just check for the ExtState and if the mouse cursor is over your window, then handle the interaction if necessary.

**Reply by sdfalk666** [reply]
I look forward to seeing what you both come up with

**Reply by helgoboss** [reply]
> Sure. I suppose another way might be for me to only set the ExtState for a single defer cycle when the item is dragged and released over an unidentified window. Then in theory you could just check for the ExtState and if the mouse cursor is over your window, then handle the interaction if necessary.

If you could do that, it would probably be safer. Otherwise it could happen that the "drop" happens later, totally out of context.

Wondering how I can test it, since I don't have NVK_Create. Maybe you can give me a small proof-of-concept script that just contains the "drag-out" code?

**Comment by nickvonkaenel** [comment]
Here's a script you could have running to randomly set the state so you can test

```lua
local file = 'FILEPATH_HERE'

function main()
    r.DeleteExtState('nvk_CREATE', 'dnd_file', false)
    if math.random() > 0.99 then
        r.SetExtState('nvk_CREATE', 'dnd_file', file, false)
    end
    r.defer(main)
end

main()
```

Then you could do something like this to accept and check for dnd

```lua
function accept_dnd_file()
    if r.HasExtState('nvk_CREATE', 'dnd_file') then
        local file = r.GetExtState('nvk_CREATE', 'dnd_file')
        r.DeleteExtState('nvk_CREATE', 'dnd_file', false) -- no other scripts can accept during the defer cycle
        return file
    end
end


if valid_dnd_bounds() then -- some kind of check that mouse in valid dnd location
    r.SetExtState('nvk_CREATE', 'dnd_valid', '1', false) -- so I can change the cursor to show its a valid drop zone. I will clear this in my script once it's checked so no need to delete it (this will mean other scripts can access it without conflicts)
    local file = accept_dnd_file()
    if file and file ~= '' then
        --handle file here
    end
end
```

**Reply by helgoboss** [reply]
Thanks, Nick! I'll give it a go. Can't use the lower part because Playtime is not a script, but I'll come up with something similar.

**Comment by sdfalk666** [comment]
I don't expect miracles (and I'm sure you're both very busy), but I thought I'd see if there's any updates on this?

thanks

Stephen

**Reply by helgoboss** [reply]
From my side, I can't work on this right away. Busy with release 2.16.4. Maybe could give it a try in 2.16.5. I'll definitely post here when I do this and you can always see what's currently in the works at https://github.com/users/helgoboss/projects/3.

**Reply by helgoboss** [reply]
Until then, there's an easy alternative: Drag into a REAPER arrangement. Then right click the item and choose "Import selected item into slot".

**Reply by sdfalk** [reply]
**Until then, there's an easy alternative: Drag into a REAPER arrangement. Then right click the item and choose "Import selected item into slot".****

Which is what I do already.
Yeah, I t's not a panic or anything, it would just be nice to have one less step when working with files.

**Comment by sdfalk** [comment]
Thought I'd see if there was any update on this

**Comment by sdfalk** [comment]
well hopefully soon

**Comment by sdfalk** [comment]
.

**Comment by helgoboss** [comment]
I will write here when there's an update. Working on new features is mostly paused until the final Playtime launch.

**Reply by sdfalk** [reply]
Thats all I warted to know.

thanks
